docs-util/packages/docblock-generator/src/classes/knowledge-base-factory.ts

```diff
@@ -0,0 +1,325 @@
+    },
+    {
+      endsWith: "Id",
+      template: (str: string): string => {
+        const formatted = camelToWords(str.replace(/Id$/, ""))
+
+        return `The ${formatted}'s ID.`
+      },
+      kind: [
+        ts.SyntaxKind.PropertySignature,
+        ts.SyntaxKind.PropertyDeclaration,
+        ts.SyntaxKind.Parameter,
+      ],
+    },
+    {
+      exact: "id",
+      template: (str, options) => {
+        if (options?.rawParentName?.startsWith("Filterable")) {
+          return `The IDs to filter the ${options?.parentName || `{name}`} by.`
+        }
+        return `The ID of the ${options?.parentName || `{name}`}.`
+      },
+      kind: [ts.SyntaxKind.PropertySignature],
+    },
+    {
+      exact: "metadata",
+      template: "Holds custom data in key-value pairs.",
+      kind: [ts.SyntaxKind.PropertySignature],
+    },
+    {
+      exact: "customHeaders",
+      template: "Custom headers to attach to the request.",
+    },
+  ]
+  private functionSummaryKnowledgeBase: KnowledgeBase[] = [
+    {
+      startsWith: "listAndCount",
+      template:
+        "retrieves a paginated list of {return type} along with the total count of available {return type} satisfying the provided filters.",
+    },
+    {
+      startsWith: "list",
+      template:
+        "retrieves a paginated list of {return type} based on optional filters and configuration.",
+    },
+    {
+      startsWith: "retrieve",
+      template: "retrieves a {return type} by its ID.",
+    },
+    {
+      startsWith: "create",
+      template: "creates a new {return type}",
+    },
+    {
+      startsWith: "delete",
+      template: "deletes {return type} by its ID.",
+    },
+    {
+      startsWith: "update",
+      template: "updates existing {return type}.",
+    },
+    {
+      startsWith: "softDelete",
+      template: "soft deletes {return type} by their IDs.",
+    },
+    {
+      startsWith: "restore",
+      template: "restores soft deleted {return type} by their IDs.",
+    },
+  ]
+  private exampleCodeBlockLine = `${DOCBLOCK_DOUBLE_LINES}\`\`\`ts${DOCBLOCK_NEW_LINE}{example-code}${DOCBLOCK_NEW_LINE}\`\`\`${DOCBLOCK_DOUBLE_LINES}`
+  private examplesKnowledgeBase: KnowledgeBase[] = [
+    {
+      startsWith: "list",
+      template: `To retrieve a list of {type name} using their IDs: ${this.exampleCodeBlockLine}To specify relations that should be retrieved within the {type name}: ${this.exampleCodeBlockLine}By default, only the first \`{default limit}\` records are retrieved. You can control pagination by specifying the \`skip\` and \`take\` properties of the \`config\` parameter: ${this.exampleCodeBlockLine}`,
+    },
+    {
+      startsWith: "retrieve",
+      template: `A simple example that retrieves a {type name} by its ID: ${this.exampleCodeBlockLine}To specify relations that should be retrieved: ${this.exampleCodeBlockLine}`,
+    },
+  ]
+  private functionReturnKnowledgeBase: KnowledgeBase[] = [
```

Comment:
nit: pluralize `{return type}` with `(s)` as we will always support operations in bulk moving forward